text/0000-guaranteed-slice-repr.md

```diff
@@ -0,0 +1,205 @@
+
+* `data` must point to `len` consecutive properly initialized values of type `T`.
+
+* The total size `len * mem::size_of::<T>()` of the slice must be no larger than `isize::MAX`,
+  and adding that size to `data` must not "wrap around" the address space.
+  See the safety documentation of [`pointer::offset`].
+
+## `str`
+
+The layout of `&str` is the same as that of `&[u8]`, and the layout of
+`&mut str` is the same as that of `&mut [u8]`. More generally, `str` behaves like
+`#[repr(transparent)] struct str([u8]);`. Safe Rust functions may assume that
+`str` holds valid UTF8, but [it is not immediate undefined-behavior to store
+non-UTF8 data in `str`](https://doc.rust-lang.org/std/primitive.str.html#invariant).
+
+## Pointers
+
+Raw pointers to slices such as `*const [T]` or `*mut str` use the same layout
+as slice references, but do not necessarily point to anything.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Zero-sized types
+
+One could imagine representing `&[T]` as only `len` for zero-sized `T`.
+This proposal would preclude that choice in favor of a standard representation
+for slices regardless of the underlying type.
+
+Alternatively, we could choose to guarantee that the data pointer is present if
+and only if `size_of::<T> != 0`. This has the possibility of breaking exising
+code which smuggles pointers through the `data` value in `from_raw_parts` /
+`into_raw_parts`.
+
+## Uninhabited types
+
+Similarly, we could be *extra* tricky and make `&[!]` or other `&[Uninhabited]`
+types into a ZST since the slice can only ever be length zero.
+
+If we want to maintain the pointer field, we could also make `&[!]` *just* a
+pointer since we know the length can only be zero.
+
+Either option may offer modest performance benefits for highly generic code
+which happens to create empty slices of uninhabited types, but this is unlikely
+to be worth the cost of maintaining a special case.
+
+## Compatibility with C++ `std::span`
+
+The largest drawback of this layout and set of validity requirements is that it
+may preclude `&[T]` from being representationally equivalent to C++'s
+`std::span<T, std::dynamic_extent>`.
+
+* `std::span` does not currently guarantee its layout. In practice, pointer + length
+  is the common representation. This is even observable using `is_layout_compatible`
+  [on MSVC](https://godbolt.org/z/Y8ardrshY), though not
+  [on GCC](https://godbolt.org/z/s4v4xehnG) nor
+  [on Clang](https://godbolt.org/z/qsd1K5oGq). Future changes to guarantee a
+  different layout in the C++ standard (unlikely due to MSVC ABI stabilitiy
+  requirements) could preclude matching the layout with `&[T]`.
+
+* Unlike Rust, `std::span` allows the `data` pointer to be `nullptr`. One
+  possibile workaround for this would be to guarantee that `Option<&[T]>` uses
+  `data: std::ptr::null(), len: 0` to represent the `None` case, making
+  `std::span<T>` equivalent to `Option<&[T]>` for non-zero-sized types.
```

Comment:
Maybe it would be worth a perf run to see if there's any measurable impact of doing this.

Reply:
One possible downside is that even guaranteeing `zeroed::<Option<&[T]>>() == None` would probably preclude using the length as the niche in `&[T]` (for non-ZST `T`[^1]) as a future possibility, though. 

[^1]: since the length can be at most `isize::MAX as usize / size_of::<T>()` due to the maximum allocation size being `isize::MAX`

Reply:
I’m not sure if the compiler internals are equipped to do that, but in theory it’s possible to use the not-null niche for `Option<&[T]>` while using the length niche for other enums that need more niche values. This would probably give marginally better codegen on most targets than using `len = -1` as representation for `None` because zero is usually the cheapest constant to materialize and test against. (And it’s definitely better than `isize::MAX + 1`, which requires several instructions to materialize on some targets.)